fix: expand k8s pod discovery regex

kruskall opened the pull request:
update k8s pod ID discovery regex to the current status of the spec.

See https://github.com/elastic/apm/blob/main/specs/agents/metadata.md#containerkubernetes-metadata

Add a testcase to make sure the pod mentioned in the related issue is
discovered correctly.

Closes https://github.com/elastic/apm-agent-go/issues/822

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1288/detail/PR-1288/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1288/detail/PR-1288/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1288/detail/PR-1288/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1288/detail/PR-1288/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-go_1288.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-08-06T23:26:29.258Z&rangeTo=2022-08-06T23:46:29.258Z&transactionName=BUILD+apm-agent-go%2Fapm-agent-go-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=719cc719f5da0d8b9d052c5753bb3096&transactionId=c29b8f18131ade95)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-08-06T23:36:29.258+0000


* Duration: 19 min 3 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 2964  |
| Skipped      | 81 |
| Total        | 3045   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![9](https://img.shields.io/badge/9%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-1288/detail/PR-1288/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 17 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1288/runs/1/steps/157/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 1 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1288/runs/1/steps/164/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 1 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1288/runs/1/steps/171/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 20 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1288/runs/1/steps/210/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 5 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1288/runs/1/steps/241/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 5 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1288/runs/1/steps/262/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 3 min 23 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1288/runs/1/steps/270/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 3 min 32 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1288/runs/1/steps/287/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 3 min 33 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-1288/runs/1/steps/315/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `run benchmark tests` : Run the benchmark test.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

axw commented on internal/apmhostutil/container_linux.go:
```diff
@@ -46,12 +46,12 @@ var (
 
 	containerIDRegexp = regexp.MustCompile(
 		"^" +
-			"[[:xdigit:]]{64}|" +
+			"[[:xdigit:]]{64}$|" +
 			"[[:xdigit:]]{8}-[[:xdigit:]]{4}-[[:xdigit:]]{4}-[[:xdigit:]]{4}-[[:xdigit:]]{4,}" +
 			"$",
```
I think the original intention was that the "^" and "$" were to apply regardless of which of the other lines matched. Possibly there was originally a non-capturing group surrounding those options. Anyway, I think we should add "^" and "$" to each of the lines, so it's:

```go
	containerIDRegexp = regexp.MustCompile(
			"^[[:xdigit:]]{64}$|" +
			"^[[:xdigit:]]{8}-[[:xdigit:]]{4}-[[:xdigit:]]{4}-[[:xdigit:]]{4}-[[:xdigit:]]{4,}$",
	)
```

WDYT?

kruskall replied:
I agree, this should also be more in line with the spec 👍

axw reviewed (commented):
Should we also add the test case from the description of https://github.com/elastic/apm/pull/508?

apmmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`59/59`) | :green_heart: 
Files | 99.346% (`152/153`) | :+1: 
Classes | 96.275% (`336/349`) | :+1: 
Methods | 90.198% (`957/1061`) | :+1: 
Lines | 82.194% (`11120/13529`) | :-1: -0.033
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

kruskall commented:
It turns out that the test case in the description was not recognized as the discovery algorithm was outdated. I've updated the code, added comments and a test case 🏓

axw commented on internal/apmhostutil/container_linux.go:
```diff
@@ -99,32 +97,40 @@ func readCgroupContainerInfo(r io.Reader) (*model.Container, *model.Kubernetes,
+		// If the basename ends with ".scope", check for a hyphen and remove everything up to
+		// and including that. This allows us to match .../docker-<container-id>.scope as well
+		// as .../<container-id>.
+		if strings.HasSuffix(basename, ".scope") {
```
I think we can remove the `systemdScopeSuffix` constant?

kruskall replied:
Good catch!
